File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
// TC- O(n)
// SC- O(n)

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    bool isInBetween(TreeNode* node, long llim, long ulim){
        if(!node)
            return true;
        if(node->val>=ulim or node->val<=llim)
            return false;
        return (isInBetween(node->left, llim, node->val) and isInBetween(node->right, node->val, ulim));
    }
public:
    bool isValidBST(TreeNode* root) {
        return isInBetween(root, LONG_MIN, LONG_MAX);
    }
};

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# TC- O(n)
# SC- O(h)

class Solution:
    def isInInterval(self, node, mini, maxi) -> bool:
        if node is None:
            return True
        elif not (mini<node.val<maxi):
            return False
        return self.isInInterval(node.left, mini, node.val) and self.isInInterval(node.right, node.val, maxi)

    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.isInInterval(root, float('-inf'), float('inf'))